### PR TITLE
raftstore: propose in-memory pessimistic locks before prepare merge

### PR DESCRIPTION
**Pull request by sticnarf**
- [x] Based on https://github.com/tikv/tikv/pull/11655

### What is changed and how it works?

Refer to #11452 

Before prepare merge, we cannot propose pessimistic locks immediately because there can be write requests removing pessimistic locks proposed but not applied.

So, if there are pessimistic locks, we firstly ban all write requests and record the current propose index, which is called `prepare_merge_fence`. After the applied index reaches `prepare_merge_fence`, we can safely propose all pessimistic locks and then propose prepare merge.

### Check List 

Tests 

- Unit tests

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by youjiali1995 — COMMENTED** [review]
Can memory locks are piggybacked by `CommitMerge`?

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs]: Sorry, it's a mistake...

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs]: I change to ignore printing the log specially for a pending merge.

**Comment by sticnarf** [comment]
> Can memory locks are piggybacked by `CommitMerge`?

It's not easy.

`CommitMerge` is proposed from the source region peer which is in the same TiKV with the target region leader. The source region peer can be a follower, which does not include the in-memory pessimistic locks. So, we cannot simply let the source region collect in-memory pessimistic locks and generate a `CommitMerge` proposal with locks.

File: components/raftstore/src/store/peer.rs
Comment (youjiali1995) [reply components/raftstore/src/store/peer.rs:3630]: I get it. Locks are piggybacked in `CommitMerge` indirectly by proposing these before the `PrepareMerge`.

File: components/raftstore/src/store/peer.rs
Comment (youjiali1995) [inline components/raftstore/src/store/peer.rs:3704]: What admin requests can be useful here except `PrepareMerge`? These admin requests may forbid merge.

File: components/raftstore/src/store/peer.rs
Comment (youjiali1995) [inline components/raftstore/src/store/peer.rs:3523]: Is it a good idea to move this check after checking `merge_max_log_gap`, admin requests and `raft_entry_max_size`? And when there is a `prepare_merge_fence`, only check apply index and do not check these again.

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs:3704]: I was thinking other admin requests are just acceptable here. `PrepareMerge` can just fail in this case.

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs]: > Is it a good idea to move this check after checking `merge_max_log_gap`, admin requests and `raft_entry_max_size`?

Ah, yes. We shouldn't block write if the log gap is too large.

Now I move it after these checks.

> And when there is a `prepare_merge_fence`, only check apply index and do not check these again.

I'm not sure... It may make the code more complex. And when the applied index reaches the fence, some pessimistic locks may be deleted and the log gap may shrink too. It will be more likely to pass the size check. Considering the frequency of region merge, perhaps a second calculation is acceptable.

File: components/raftstore/src/store/peer.rs
Comment (youjiali1995) [inline components/raftstore/src/store/peer.rs:3603]:
```suggestion
        // Check passed, clear fence and start proposing PrepareMerge.
        self.prepare_merge_fence = 0;
        self.pending_prepare_merge = None;
```

File: components/raftstore/src/store/peer.rs
Comment (youjiali1995) [reply components/raftstore/src/store/peer.rs]: > And when the applied index reaches the fence, some pessimistic locks may be deleted and the log gap may shrink too. It will be more likely to pass the size check.

The size check can't block merge in this case, but is it possible there is a new admin request within the log gap that forbid to merge because admin requests can pass `prepare_merge_fence` check? We should clear fence if it happens.

Comment (youjiali1995) [reply components/raftstore/src/store/peer.rs]: Clear the fence when:
1. pass all checks and start proposing `PrepareMerge`.
2. fail to propose `PrepareMerge` and the fence exsits.

Comment (sticnarf) [reply components/raftstore/src/store/peer.rs]: > is it possible there is a new admin request within the log gap that forbid to merge because admin requests can pass `prepare_merge_fence` check

Definitely. Now, I use the similar way as before, clear the fence before prepare merge checks. Hope to reduce subtleties burning brains...

**Review by youjiali1995 — APPROVED** [review]
LGTM, but it's better to add tests to cover recent changes.

File: components/raftstore/src/store/peer.rs
Comment (5kbpers) [inline components/raftstore/src/store/peer.rs:3666]: How about using `propose_raft_command_internal` here?

File: components/raftstore/src/store/peer.rs
Comment (5kbpers) [inline components/raftstore/src/store/peer.rs:3600]: Could we assert it should always be true because all cases have already been checked?

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs:3687]: Seems not possible to call `propose_raft_command_internal` which is a function of `PeerFsmDelegate` from `Peer`?

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs:3617]: I think it's possible that it's going to have a transfer leader, so `is_valid` becomes false. And related commands also have been committed, so the checks are pased.

Comment (5kbpers) [reply components/raftstore/src/store/peer.rs:3617]: Can you add some comments for clarifying this?

File: components/raftstore/src/store/peer.rs
Comment (NingLin-P) [inline components/raftstore/src/store/peer.rs:3517]: Seems here is the only place (except leader change) that may update `prepare_merge_fence`, does it mean we very rely on PD to retry the merge command to push forward the merge process? If the merge request is keep failing (perhaps due to the `pessimistic_locks` is too large) and PD end the merge operator with a timeout error hence won't retry the merge command anymore, the region may remain in merging mode forever and won't accept incoming proposals.

Comment (sticnarf) [reply components/raftstore/src/store/peer.rs:3517]: It is also retried in `on_apply_res`. Which means, when the `applied_index` reaches `prepare_merge_fence`, we will continue.

File: components/raftstore/src/store/peer.rs
Comment (NingLin-P) [inline components/raftstore/src/store/peer.rs:3654]: Seems `pessimistic_locks.memory_size` is the sum of all `key.len() + locks.memory_size()`, although we have reserve `raft_entry_max_size * 0.1` for extra fields in `CommitMerge`, depend on the number of pessimistic locks, more extra fields will bring by `PutRequest`, will their size exceed the reserved size?

File: components/raftstore/src/store/peer.rs
Comment (NingLin-P) [inline components/raftstore/src/store/peer.rs:1672]: Perhaps we should clear them when the leader step down instead.

File: components/raftstore/src/store/peer.rs
Comment (NingLin-P) [reply components/raftstore/src/store/peer.rs:3517]: In `on_apply_res`, we retried with the previous saved `PrepareMerge` request, it is also required the request can still pass all the checks before `pre_propose_prepare_merge`

Comment (sticnarf) [reply components/raftstore/src/store/peer.rs:3517]: Sounds like something worth considering. But what check may it fail to pass before `pre_propose_prepare_merge` while there is not any region change and the region can still serve request?

Comment (NingLin-P) [reply components/raftstore/src/store/peer.rs:3517]: I have been checked the code, checks that rely on states that outside of the source peer, like the disk full check and the target peer check:
https://github.com/tikv/tikv/blob/4e2af0426c13657c6eeb69bffc7b9847f2f61b47/components/raftstore/src/store/fsm/peer.rs#L3886-L3902

which may change after waiting for the source peer's `appied_index` catch up. 

The source peer's state can also change as apply more log, but with the `epoch_checker` there won't have `EpochNotMatch` error though.

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs:3654]: Protobuf integers and bytes use variant encoding, it is hard to compare (usually the proposal will be smaller than the original map).
I will add a second check of the size.

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs:3514]: Now I always clear the fence after proposing `PrepareMerge` in `on_apply_res`. I think the region will return available even if `PrepareMerge` fails to propose.

File: components/raftstore/src/store/peer.rs
Comment (sticnarf) [reply components/raftstore/src/store/peer.rs]: Moved to `clear_in_memory_pessimistic_locks`

**Comment by sticnarf** [comment]
@NingLin-P Thanks for the comments! I've addressed them and could you take another look?

File: components/raftstore/src/store/peer.rs
Comment (NingLin-P) [inline components/raftstore/src/store/peer.rs:3646]: If the `PrepareMerge` command fails to propose later on (`raft_group.propose` may still return error), this flag may reject incoming retrying and cause the region can't be merged until leadership change or `pessimistic_locks ` become empty.

**Review by NingLin-P — COMMENTED** [review]
Rest LGTM

**Review by NingLin-P — APPROVED** [review]

**Comment by sticnarf** [comment]
I will merge it first.

I've added FIXME in apply to the last two comments. And I will address them in https://github.com/tikv/tikv/pull/11883.

**Comment by sticnarf** [comment]
> I will merge it first.
> 
> I've added FIXME in reply to the last two comments. And I will address them in #11883.

After some thoughts, the problem is not suitable to be fixed by tick. Instead, I open https://github.com/tikv/tikv/pull/11985 to fix it (unit tests to be added)

**Pull request merged**
